### ISSUE TITLE
A few questions about how the code works

### ISSUE THREAD

**Issue by brightproject** [RobTillaart/MS5611#39]
Good afternoon @RobTillaart 
I run this code for two sensors

```cpp
#include "MS5611.h"

MS5611 MS5611_1(0x76);
MS5611 MS5611_2(0x77);

uint32_t start1, stop1, start2, stop2, count;

void setup()
{
  Serial.begin(115200);
  while (!Serial);
  Serial.println();
  Serial.println(__FILE__);
  Serial.print("MS5611_LIB_VERSION: ");
  Serial.println(MS5611_LIB_VERSION);
  Serial.println();

  Wire.begin();
  if (MS5611_1.begin() == true)
  {
    Serial.println("MS5611_1 found.");
  }
  else
  {
    Serial.println("MS5611_1 not found. halt.");
    while (1);
  }
  if (MS5611_2.begin() == true)
  {
    Serial.println("MS5611_2 found.");
  }
  else
  {
    Serial.println("MS5611_2 not found. halt.");
    while (1);
  }
  count = 0;
}

void loop()
{
  delay(1000);

  // MS5611_1.setOversampling(OSR_LOW);
  MS5611_1.setOversampling(OSR_ULTRA_HIGH);
  start1 = micros();
  int result_1 = MS5611_1.read();   //  uses OSR_ULTRA_HIGH
  stop1 = micros();
  start2 = micros();
  int result_2 = MS5611_2.read();   //  uses default OSR_ULTRA_LOW  (fastest)
  stop2 = micros();

  if (count % 20 == 0)
  {
    Serial.println();
    Serial.println("CNT\tOSR1\tDUR1\tRES1\tTEMP1\tPRES1\tOSR2\tDUR2\tRES2\tTEMP2\tPRES2\tDIFF_T\tDIFF_P");
  }

  Serial.print(count);
  count++;
  Serial.print("\t");
  Serial.print( (uint8_t) MS5611_1.getOversampling());
  Serial.print("\t");
  Serial.print(stop1 - start1);
  Serial.print("\t");
  Serial.print(result_1);
  Serial.print("\t");
  Serial.print(MS5611_1.getTemperature(), 2);
  Serial.print("\t");
  Serial.print(MS5611_1.getPressure(), 2);
  Serial.print("\t");
  Serial.print( (uint8_t) MS5611_2.getOversampling());
  Serial.print("\t");
  Serial.print(stop2 - start2);
  Serial.print("\t");
  Serial.print(result_2);
  Serial.print("\t");
  Serial.print(MS5611_2.getTemperature(), 2);
  Serial.print("\t");
  Serial.print(MS5611_2.getPressure(), 2);
  Serial.print("\tT: ");
  Serial.print(MS5611_1.getTemperature() - MS5611_2.getTemperature(), 2);
  Serial.print("\tP: ");
  Serial.print(MS5611_1.getPressure() - MS5611_2.getPressure(), 2);
  Serial.println();
}
```
I get the following output in the serial port

```
src\main.cpp
MS5611_LIB_VERSION: 0.4.0

MS5611_1 found.
MS5611_2 found.

CNT     OSR1    DUR1    RES1    TEMP1   PRES1   OSR2    DUR2    RES2    TEMP2   PRES2   DIFF_T  DIFF_P
0       12      19763   0       28.26   1009.77 8       2760    0       28.11   1009.29 T: 0.15 P: 0.48
1       12      19763   0       28.26   1009.76 8       2760    0       28.11   1009.41 T: 0.15 P: 0.35
2       12      19763   0       28.27   1009.76 8       2760    0       28.13   1009.33 T: 0.14 P: 0.43
```
I have a simple interest in the values ​​in the columns `RES1` and `RES2`.
I also don't understand why different `OSR's` give approximately the same data from each sensor, although the resolution should be different.
Previously worked with `BMP280` sensors and when changing the `OSR`, the accuracy of the sensors changes.
I will be glad to start a discussion with you🙂

(updated syntax highlighting)

**Comment by RobTillaart** [comment]
Thanks for your questions, 
Might take a few days to dive into it.
R.

**Comment by RobTillaart** [comment]
With sensor 1 you should notice that the readings are the same in the 0.01 range while sensor two varies (roughly) 0.1 between measurements. 
Although the absolute difference looks small it still is a factor 10.

An experiment you could do is to "whistle" over the sensors so the get variation in air  pressure. I expect you should see a bigger delta between the readings of both sensors.

What you should also notice is the difference in duration of the calls. The extra digit in accuracy takes about 6x as much time, so depending on your requirements you choose the oversampling level.

**Comment by brightproject** [comment]
> Although the absolute difference looks small it still is a factor 10.

It seems to me that you are using incorrect discrediting of `ADC` sensors by entering `ODR` coefficients.
You need to use the bits from the datasheet to set the sensor data sampling frequencies.
I believe that `ODR` will be different for `SPI` and `I2C`.

**Comment by RobTillaart** [comment]
I recall the SPI version had problems with self heating effect. 

Need to reread the datasheet in detail to understand what you mean exactly. Do you have a reference or proposal how to improve the code?

**Comment by brightproject** [comment]
> Do you have a reference or proposal how to improve the code?

I'm making an air data computer device.
I think there will be test parameters soon.

![baro_sensor](https://github.com/user-attachments/assets/a6f8ab4b-5cb3-4fcc-95d9-3810a5717457)

![5226480234772164764](https://github.com/user-attachments/assets/ae4dc0e1-b383-4da2-af46-3f8d5026eefe)

**Comment by RobTillaart** [comment]
Nice box, looks pro, what's the application?

**Comment by brightproject** [comment]
> Nice box, looks pro, what's the application?

Homemade sensor measuring differential gas pressure.

**Comment by brightproject** [comment]
I still don't understand the library's output.
Here's an example of the code

> int result = MS5611.read();

It returns `0`, what does this mean - `no errors` or how should I understand it?

**Comment by RobTillaart** [comment]
Best way to learn to understand the cose is to read it carefully.

The library read() fetches data from the device, and the other functions give access to the fetched data.

**Comment by RobTillaart** [comment]
The return value 0 is indeed OK.

**Comment by brightproject** [comment]
I understand that, thanks @RobTillaart for the clarification🙂
I am also interested in the function

> https://github.com/RobTillaart/MS5611/blob/09317704722ef7e40b40b97de2444fbd56cb3d64/MS5611.cpp#L175C1-L179C3

The pressure is in millibars, right?
I just need to get the pressure in `Pascals`, and as I understand it, just don't multiply by `_pressure * 0.01`
I am interested in your [other](https://github.com/RobTillaart/pressure) library, but I would not like to use it in this project, because I do not need so many different pressure units, and I just want to add a couple of functions to my code.

I also tried using these two functions
```cpp
// EXPERIMENTAL
uint16_t MS5611::getManufacturer()
{
return readProm(0);
}

// EXPERIMENTAL
uint16_t MS5611::getSerialCode()
{
return readProm(7) >> 4;
}
```
But I got `ZERO` as output.
Is it possible that my sensor is fake?

**Comment by RobTillaart** [comment]
1 milliBar = 100 Pascal.  
1 Bar = 100000 Pascal

To convert millibar to Pascal you have to multiply by 100.0

e.g. 1023.71 mBar = 102371 Pascal

include this function (depending on you wanting to have float or integer type.

```cpp
float mBar2Pascal(float mbar) { return mbar * 100; };
or
int32_t  mBar2Pascal(float mbar) { return round(mbar * 100); };  // (delta) pressure can be negative
```

BTW if you use the pressure library the linker will normally only include those parts of the code that are used.

**Comment by RobTillaart** [comment]
The two functions are based upon an image from the datasheet.

![image](https://github.com/user-attachments/assets/d98e0264-914c-455f-96c8-3e56c2c114ce)

As one was dedicated to manufacturer **getManufacturer()** seems to be a reasonable name to access it.
SerialCode I do not recall the why, it might be mentioned in some related datasheet (similar sensor)

Note: typo in the readme.md. PROM[6] should be PROM[7] for getSerialCode()

----

FYI, 
I have 6 PROM dumps with different values for the **getManufacturer()** private data and 0 is one of them.
See discussion in #33 

Those PROM dumps do not include the PROM[7] values unfortunatelly.

**Comment by RobTillaart** [comment]
It is possible to add **getPressurePascal()** as Pascal is the SI unit.

```cpp
float getPressurePascal(); //  returns pressure in Pascal.
```

**Comment by RobTillaart** [comment]
@brightproject 

Created a develop branch that includes **getPressurePascal()** + two developer functions.
If possible, please verify if **getPressurePascal()** works for you.

**Comment by RobTillaart** [comment]
@brightproject 
Merged develop branch into master and released 0.4.1

You should now be able to call **getPressurePascal()**

Thanks again for the issue, appreciated.
